### PR TITLE
Any code fixes

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Fixes a bunch of issues in the AnyCode implementation

### CODE REVIEW

File: Nodejs/Product/Npm/SPI/NpmCommand.cs [hunk 1]
```diff
@@ -106,7 +106,7 @@ public void CancelCurrentTask()
                 cancelled = true;
             }
             OnCommandCompleted(this.Arguments, redirector?.HasErrors ?? false, cancelled);
-            return !redirector.HasErrors;
+            return redirector == null || !redirector.HasErrors; // return true when the command completed without errors, or redirector is null (which means we can't track)
```
Comment (amcasey) [inline Nodejs/Product/Npm/SPI/NpmCommand.cs]: >redirector == null || !redirector.HasErrors [](start = 19, length = 43)

Alternatively, `redirectory?.HasErrors != true`.

File: Nodejs/Product/Nodejs/SharedProject/OutputWindowWrapper.cs [hunk 2]
```diff
@@ -24,41 +24,37 @@ public sealed class OutputPaneWrapper
...
-            if (this.lazyOutputPaneCollection.TryGetValue(target, out var lazyOutputPane))
-            {
-                return lazyOutputPane;
-            }
-            else
-            {
-                ThreadHelper.ThrowIfNotOnUIThread();
-                var (title, guid) = GetPaneInfo(target);
-                var outputWindow = (IVsOutputWindow)this.ServiceProvider.GetService(typeof(SVsOutputWindow));
+            //...
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/OutputWindowWrapper.cs]: >//... [](start = 12, length = 5)

Why is this here?

File: Nodejs/Product/Nodejs/SharedProject/OutputWindowWrapper.cs [hunk 3]
```diff
@@ -24,41 +24,37 @@ public sealed class OutputPaneWrapper
...
+                    return;
                 }
 
-                this.lazyOutputPaneCollection.Add(target, lazyOutputPane);
-                return lazyOutputPane;
+                // Must activate the workspace pane for it to show up in the output window
+                lazyOutputPane.Activate();
             }
 
+            this.lazyOutputPaneCollection.Add(target, lazyOutputPane);
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/SharedProject/OutputWindowWrapper.cs]: >this.lazyOutputPaneCollection.Add(target, lazyOutputPane); [](start = 12, length = 58)

If this is the only place where this method mutates `this`, it might be cleaner to return the result and have the caller update the map.  In fact, rather than pass `target` in at all, why not inline the results of `GetPaneInfo` at the only two call sites?

**Comment by amcasey** [comment]
>                 var fileContextActions = await node.Workspace.GetFileContextActionsAsync(packageJson.Main, new[] { DebugLaunchActionContext.ContextTypeGuid });

`workspace`

---
Refers to: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs:170 in f0a4177. [](commit_id = f0a417717f734f7cbafd44b6bf32acc9eecef01b, deletion_comment = False)

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs [hunk 4]
```diff
@@ -116,44 +117,43 @@ public int Exec(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uin
...
-                using (var npmController = this.CreateController(node.Workspace))
+                using (var npmController = this.CreateController(node.FullPath))
                 using (var commander = npmController.CreateNpmCommander())
                 {
                     await commander.Install();
                 }
             }
 
-            private void ExecNpmInstallNew(WorkspaceVisualNodeBase node)
+            private void ExecNpmInstallNew(IFileNode node)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: >IFileNode [](start = 43, length = 9)

Does this fix something or is it just nice to have?

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs [hunk 5]
```diff
@@ -265,9 +265,11 @@ private static bool TryGetCommand(uint nCmdID, string filePath, out string comma
                 return false;
             }
 
-            private INpmController CreateController(IWorkspace workspace)
+            private INpmController CreateController(string packageJsonPath)
             {
-                var projectHome = workspace.Location;
+                Debug.Assert(Path.IsPathRooted(packageJsonPath) && PackageJsonFactory.IsPackageJsonFile(packageJsonPath));
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: >&& [](start = 64, length = 2)

Personally, I would split this into two asserts.

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs [hunk 6]
```diff
@@ -116,44 +117,43 @@ public int Exec(List<WorkspaceVisualNodeBase> selection, Guid pguidCmdGroup, uin
             // Note: all the Exec commands are async, this allows us to call them in a fire and forget
             // pattern, without blocking the UI or losing any logging
 
-            private async void ExecNpmInstallMissing(WorkspaceVisualNodeBase node)
+            private async void ExecNpmInstallMissing(IFileNode node)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: >IFileNode [](start = 53, length = 9)

Why pass the `IFileNode`, rather than the path?

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs [hunk 7]
```diff
@@ -4,16 +4,19 @@
...
 using Microsoft.NodejsTools.Npm;
 using Microsoft.NodejsTools.NpmUI;
 using Microsoft.VisualStudio;
 using Microsoft.VisualStudio.OLE.Interop;
+using Microsoft.VisualStudio.Shell;
 using Microsoft.VisualStudio.Workspace;
 using Microsoft.VisualStudio.Workspace.Debug;
 using Microsoft.VisualStudio.Workspace.VSIntegration.UI;
 using Microsoft.VisualStudioTools.Project;
+using ShellInterop = Microsoft.VisualStudio.Shell.Interop;
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: >ShellInterop [](start = 6, length = 12)

Where is this used?

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs [hunk 8]
```diff
@@ -166,6 +168,11 @@ private void ExecNpmInstallNew(WorkspaceVisualNodeBase node)
                 var workspace = node.Workspace;
                 var packageJson = PackageJsonFactory.Create(((IFileNode)node).FullPath);
 
+                if (string.IsNullOrEmpty(packageJson.Main))
+                {
+                    return;
+                }
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: Does this mean the user's request is silently ignored?

File: Nodejs/Product/Nodejs/Workspace/PackageJsonScannerFactory.cs [hunk 9]
```diff
@@ -43,13 +43,16 @@ protected override Task<List<FileReferenceInfo>> ComputeFileReferencesAsync(stri
 
                 var packageJson = PackageJsonFactory.Create(filePath);
                 var main = packageJson.Main;
-                var fileReferences = new List<FileReferenceInfo>
+
+                var fileReferences = new List<FileReferenceInfo>();
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Workspace/PackageJsonScannerFactory.cs]: >new List<FileReferenceInfo>(); [](start = 37, length = 30)

Seems like it might be (slightly) preferable to return `Array.Empty` if there's no `main`.

File: Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs (hunk 8 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Workspace/ContextMenuProvider.cs]: Yes, there was no good way to show a dialog. I'm talking to AnyCode to see how we can hide the menu item
